nomad/acl.go

```diff
@@ -1,16 +1,128 @@
 package nomad
 
 import (
+	"errors"
 	"fmt"
+	"net"
 	"time"
 
 	metrics "github.com/armon/go-metrics"
 	lru "github.com/hashicorp/golang-lru"
 	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/helper"
 	"github.com/hashicorp/nomad/nomad/state"
 	"github.com/hashicorp/nomad/nomad/structs"
 )
 
+// Authenticate extracts an AuthenticatedIdentity from the request context or
+// provided token. The caller can extract an acl.ACL, WorkloadIdentity, or other
+// identifying token to use for authorization.
+//
+// Note: when called on the follower we'll be making stale queries, so it's
+// possible if the follower is behind that the leader will get a different value
+// if an ACL token or allocation's WI has just been created.
+func (s *Server) Authenticate(ctx *RPCContext, secretID string) (*structs.AuthenticatedIdentity, error) {
+	// get the user ACLToken or anonymous token
+	aclToken, err := s.ResolveSecretToken(secretID)
+	identity := &structs.AuthenticatedIdentity{ACLToken: aclToken}
```

Comment:
Ok, I've done that throughout the first half of the function. For everything after that, where we reach the server identification through the context, it gets noticeably less readable because you've got to get those non-disjoint pieces of data together.